examples/stable_fluid.py

```diff
@@ -95,83 +153,87 @@ def apply_impulse(vf: ti.template(), dyef: ti.template(),
         dx, dy = (i + 0.5 - omx), (j + 0.5 - omy)
         d2 = dx * dx + dy * dy
         # dv = F * dt
-        factor = ti.exp(-d2 * inv_force_radius)
-        momentum = mdir * f_strength_dt * factor
+        factor = ti.exp(-d2 / force_radius)
+        momentum = mdir * f_strength * dt * factor
         v = vf[i, j]
         vf[i, j] = v + momentum
         # add dye
         dc = dyef[i, j]
         if mdir.norm() > 0.5:
-            dc += ti.exp(-d2 * inv_dye_denom) * ti.Vector(
+            dc += ti.exp(-d2 * (4 / (res / 15)**2)) * ti.Vector(
                 [imp_data[4], imp_data[5], imp_data[6]])
         dc *= dye_decay
         dyef[i, j] = dc
 
 
 @ti.kernel
 def divergence(vf: ti.template()):
+    ti.cache_read_only(vf)
     for i, j in vf:
-        vl = sample(vf, i - 1, j)[0]
-        vr = sample(vf, i + 1, j)[0]
-        vb = sample(vf, i, j - 1)[1]
-        vt = sample(vf, i, j + 1)[1]
+        vl = sample(vf, i - 1, j).x
+        vr = sample(vf, i + 1, j).x
+        vb = sample(vf, i, j - 1).y
+        vt = sample(vf, i, j + 1).y
         vc = sample(vf, i, j)
         if i == 0:
-            vl = -vc[0]
+            vl = -vc.x
         if i == res - 1:
-            vr = -vc[0]
+            vr = -vc.x
         if j == 0:
-            vb = -vc[1]
+            vb = -vc.y
         if j == res - 1:
-            vt = -vc[1]
-        velocity_divs[i, j] = (vr - vl + vt - vb) * half_inv_dx
+            vt = -vc.y
+        velocity_divs[i, j] = (vr - vl + vt - vb) * 0.5
 
 
-p_alpha = -dx * dx
+@ti.kernel
+def vorticity(vf: ti.template()):
+    ti.cache_read_only(vf)
+    for i, j in vf:
+        vl = sample(vf, i - 1, j).y
+        vr = sample(vf, i + 1, j).y
+        vb = sample(vf, i, j - 1).x
+        vt = sample(vf, i, j + 1).x
+        vc = sample(vf, i, j)
+        velocity_curls[i, j] = (vr - vl - vt + vb) * 0.5
 
 
 @ti.kernel
 def pressure_jacobi(pf: ti.template(), new_pf: ti.template()):
+    ti.cache_read_only(pf)
```

Comment:
Btw, can we document this feature somewhere? I'm puzzled, despite it brings a bit of FPS improvements, and I guess I'm using it correct...